### DISCUSSION TITLE
How do we set up filesystem error logging to Application Insights?

### DISCUSSION THREAD

**Discussion by denious** [Azure/azure-functions-sql-extension#749] [General]
We are having issues logging filesystem errors to Application Insights. When streaming filesystem logs, we clearly see errors in the output, but we have not found a way to monitor them.

We are also not seeing the failed invocation count go up when errors are reported in filesystem logs. The invocation success count is going up, but not the error count.

To summarize, when things work well, all the successes are logged to Application Insights, but not the errors. What can we do to solve that?

**Comment by Charles-Gagnon** [comment]
What kinds of errors are you seeing/expect to see as a failed invocation? SqlClient errors or something else?

And then are you using SQL input/output bindings, SQL trigger bindings, or both?

**Reply by denious** [reply]
SQL trigger.

I would expect all errors to result in a failed invocation and for all errors to be reported to Application Insights. For example, the "Invalid operation. The connection is closed." error, for which there's an open issue, only appears to be visible when streaming filesystem logs, but does not get logged anywhere beyond that.

**Reply by denious** [reply]
Looks like all filesystem are going to the FunctionAppLogs diagnostic logs category and we are now able to capture most of them in Application Insights after adding the Microsoft.Azure.WebJobs.Logging.ApplicationInsights package and tweaking the host.json file.

**Reply by Charles-Gagnon** [reply]
Thanks for following up, could you share your host.json configuration? I'll take a look at what we have and make sure it makes sense - and update our docs to clarify this as well.

**Reply by denious** [reply]
**Required Package**
- Microsoft.Azure.WebJobs.Logging.ApplicationInsights

**Diagnostics Setup**
![image](https://user-images.githubusercontent.com/7120854/226681018-f5b600f1-f441-4901-9e09-dba903ea4661.png)

**host.json**
```json
{
  "logging": {
    "applicationInsights": {
      "snapshotConfiguration": {
        "handleUntrackedExceptions": true,
        "isEnabled": true,
        "isEnabledInDeveloperMode": false,
        "isEnabledWhenProfiling": true
      },
      "samplingSettings": {
        "isEnabled": true,
        "excludedTypes": "Request;Exception;Event;Dependency",
        "maxTelemetryItemsPerSecond": 20
      }
    },
    "enableDependencyTracking": true,
    "dependencyTrackingOptions": {
      "enableSqlCommandTextInstrumentation": true
    },
    "logLevel": {
      "default": "Information"
    }
  },
  "version": "2.0"
}
```

**Code Gotchas**
- Do not add .AddLogging() to FunctionsStartup
- Do not add TelemetryClient to the DI container

**Reply by Charles-Gagnon** [reply]
From looking at that package it seems like it's mostly intended for custom telemetry options. You should be able to view the logs directly in app insights if you set it up as described here : https://learn.microsoft.com/en-us/azure/azure-functions/configure-monitoring?tabs=v2#enable-application-insights-integration

Is there something you're doing that makes it so using this package is a requirement for your scenario?

**Reply by Charles-Gagnon** [reply]
Oh ok, so looking at it you're trying to send it to a Log Analytics workspace which apparently does require the extra setup you found : https://learn.microsoft.com/en-us/azure/azure-functions/functions-monitor-log-analytics?tabs=csharp#setting-up

For my own understanding - is there something making you want to send it to the workspace directly like this? From what I can tell App Insights can be configured to use a Log Analytics workspace so it might be easier to do that instead since that's something that's available right out of the box when creating a function app.

**Reply by denious** [reply]
> From looking at that package it seems like it's mostly intended for custom telemetry options. You should be able to view the logs directly in app insights if you set it up as described here : https://learn.microsoft.com/en-us/azure/azure-functions/configure-monitoring?tabs=v2#enable-application-insights-integration
> 
> Is there something you're doing that makes it so using this package is a requirement for your scenario?

We are outputting custom events to Application Insights.

> Oh ok, so looking at it you're trying to send it to a Log Analytics workspace which apparently does require the extra setup you found : https://learn.microsoft.com/en-us/azure/azure-functions/functions-monitor-log-analytics?tabs=csharp#setting-up
> 
> For my own understanding - is there something making you want to send it to the workspace directly like this? From what I can tell App Insights can be configured to use a Log Analytics workspace so it might be easier to do that instead since that's something that's available right out of the box when creating a function app.

Can you elaborate on "it might be easier to do that instead since that's something that's available right out of the box when creating a function app"? The current setup is piping only filesystem logs to the workspace, everything else does live in a separate Application Insights resource.

Should disabling the workspace integration result in filesystem logs going to that same Application Insights resource?

**Reply by Charles-Gagnon** [reply]
I'll admit this is an area I'm not deeply familiar with myself so there might be some confusion on my part. But just to make sure I'm not missing something here, the setup I was referring to was this option when creating the function app:

![image](https://user-images.githubusercontent.com/28519865/228060939-5b819bfa-a807-4d82-9c7d-393b6449a40c.png)

When I do this then I get the application logs going to appinsights and can query them along with all the other logs (under the traces table)

![image](https://user-images.githubusercontent.com/28519865/228067942-579eb738-ff44-4883-be3f-e353eee12c14.png)

So I think this just seems to be that there's multiple ways to direct where logs end up (and from what I've seen about people asking similar questions it seems like this is a common thing people run in to and has changed much over time). 

If we think there's anything specific to our function and how we do logging that can be improved I'm glad to document it here - but for general stuff like how to set up logging I'd rather avoid duplicating anything and instead just rely on people using the official Azure Functions docs for getting those questions answered. Does that sound reasonable?
